File: app/src/main/java/no/nav/tilbakemeldingsmottak/util/OidcUtils.java
package no.nav.tilbakemeldingsmottak.util;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.security.token.support.core.context.TokenValidationContextHolder;
import no.nav.security.token.support.core.jwt.JwtToken;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;

@Component
@Slf4j
@RequiredArgsConstructor
public class OidcUtils {
    private final TokenValidationContextHolder tokenValidationContextHolder;

    public Optional<String> getSubjectForIssuer(String issuer) {
        JwtToken userToken = tokenValidationContextHolder.getTokenValidationContext().getJwtToken(issuer);
        if (userToken == null) {
            return Optional.empty();
        } else {
            try {
                return Optional.of(userToken.getSubject());
            } catch (Exception e) {
                throw new RuntimeException("Feil i parsing av token",e);
            }
        }
    }

    public Optional<String> getEmailForIssuer(String issuer) {
        JwtToken userToken = tokenValidationContextHolder.getTokenValidationContext().getJwtToken(issuer);
        if (userToken == null) {
            return Optional.empty();
        } else {
            try {
                return Optional.of(userToken.getJwtTokenClaims().getStringClaim("upn"));
            } catch (Exception e) {
                throw new RuntimeException("Feil i parsing av token",e);
            }
        }
    }

    public String getFirstValidToken() {
        return tokenValidationContextHolder.getTokenValidationContext().getFirstValidToken().map(JwtToken::getTokenAsString)
                .orElseThrow(() -> new RuntimeException("Finner ikke validert OIDC-token"));
    }

    public Optional<String> getSubjectForFirstValidToken() {
        Optional<JwtToken> userToken =  tokenValidationContextHolder.getTokenValidationContext().getFirstValidToken();
        if (!userToken.isPresent()) {
            return Optional.empty();
        } else {
            try {
                return Optional.of(userToken.get().getSubject());
            } catch (Exception e) {
                return Optional.empty();
            }
        }
    }

    public void checkSubjectOnEachIssuer() {
        List<String> issuers = tokenValidationContextHolder.getTokenValidationContext().getIssuers();
        issuers.stream()
                .forEach(e -> logClaims(e));
    }


    private void logClaims(String issuer) {
        Optional<JwtToken> userToken = tokenValidationContextHolder.getTokenValidationContext().getJwtTokenAsOptional(issuer);
        log.info("Context issuer=" + issuer + " User="+ (userToken.isPresent()? userToken.get().getSubject(): "Ikke funnet"));
    }
}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/util/OidcUtils.kt
package no.nav.tilbakemeldingsmottak.util

import org.slf4j.LoggerFactory
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.stereotype.Component

@Component
class OidcUtils {

    private val log = LoggerFactory.getLogger(OidcUtils::class.java)

    fun currentJwt(): Jwt? {
        val auth = SecurityContextHolder.getContext().authentication
        return (auth as? JwtAuthenticationToken)?.token
    }

    fun getSubject(): String? =
        currentJwt()?.subject

    fun getEmail(): String? =
        currentJwt()?.claims?.get("preferred_username") as? String

    fun getPid(): String? =
        currentJwt()?.claims?.get("pid") as? String

    fun getIssuer(): String? =
        currentJwt()?.claims?.get("iss") as? String

    fun getClaim(claimName: String): Any? =
        currentJwt()?.claims?.get(claimName)

    fun isLoggedIn(isInternal: Boolean): Boolean =
        if (getPid() != null)
            true
        else
            false

    fun logClaims() {
        val jwt = currentJwt()
        if (jwt == null) {
            log.info("No JWT in security context")
        } else {
            log.info("JWT issuer=${jwt.issuer}, subject=${jwt.subject}, claims=${jwt.claims.keys}")
        }
    }

    /**
     * Returns the raw Bearer token string (for forwarding to downstream services if necessary).
     */
    fun getRawTokenValue(): String? =
        currentJwt()?.tokenValue
}
